### DISCUSSION TITLE
Any interest in proposing SQL Parser as a separate library?

### DISCUSSION THREAD

**Discussion by hungtcs** [ClickHouse/ClickHouse#60267] [Ideas]
I'm writing a tool that parses SQL and modifies it via AST using golang, that need to be compatible with postgres and clickhouse sql syntax.

I've been searching for a library that parses ClickHouse SQL syntax for a long time now, but I'm not having much luck.

I've tried some of the projects posted above GitHub, like [AfterShip/clickhouse-sql-parser](https://github.com/AfterShip/clickhouse-sql-parser), but they have a long way to go on the syntax compatibility level.

For the postgres database, I found [pganalyze/pg_query_go](https://github.com/pganalyze/pg_query_go), which is a go-language binding that itself is written in C, which is a near-perfect solution to my need to parse the postgres sql syntax.

But for the clickhouse database, I can hardly find a perfectly usable sql parser for the go language, I found that Clickhouse's sql parsing syntax code is located at <https://github.com/ClickHouse/ClickHouse/tree/master/src/Parsers> , Imagine if you refer to [pganalyze/pg_query_go](https://github.com/pganalyze/pg_query_go) and present the code as a separate library and add go bindings, can you also parse Clickhouse sql with go code?

**Comment by vdimir** [comment]
I assume one of the options to achieve this is just to add support on the ClickHouse side (good task, but it needs to be considered how to done it property), not to support external libraries.
In fact, we already support the `EXPLAIN AST` query that prints the tree in some format. However, it's impossible to convert it back to the query. But for some use cases, `EXPLAIN AST`, `EXPLAIN SYNTAX`, and `EXPLAIN QUERY TREE` may be enough.

```sql
EXPLAIN AST
SELECT
    *,
    status,
    formatReadableTimeDelta(loading_duration) AS lt,
    formatReadableSize(bytes_allocated) AS mem,
    concat(formatReadableTimeDelta(now() - last_successful_update_time), ' ago') AS last_update
FROM system.dictionaries


┌─explain──────────────────────────────────────────────────────┐
│ SelectWithUnionQuery (children 1)                            │
│  ExpressionList (children 1)                                 │
│   SelectQuery (children 2)                                   │
│    ExpressionList (children 5)                               │
│     Asterisk                                                 │
│     Identifier status                                        │
│     Function formatReadableTimeDelta (alias lt) (children 1) │
│      ExpressionList (children 1)                             │
│       Identifier loading_duration                            │
│     Function formatReadableSize (alias mem) (children 1)     │
│      ExpressionList (children 1)                             │
│       Identifier bytes_allocated                             │
│     Function concat (alias last_update) (children 1)         │
│      ExpressionList (children 2)                             │
│       Function formatReadableTimeDelta (children 1)          │
│        ExpressionList (children 1)                           │
│         Function minus (children 1)                          │
│          ExpressionList (children 2)                         │
│           Function now (children 1)                          │
│            ExpressionList                                    │
│           Identifier last_successful_update_time             │
│       Literal ' ago'                                         │
│    TablesInSelectQuery (children 1)                          │
│     TablesInSelectQueryElement (children 1)                  │
│      TableExpression (children 1)                            │
│       TableIdentifier system.dictionaries                    │
└──────────────────────────────────────────────────────────────┘
```

**Comment by hungtcs** [comment]
One of the reasons for this idea is because I have found that when I execute SQL in the clickhouse cli, it will reformat and output the SQL I executed.

Combining the code in examples, so I'm sure clickhouse already has a way to parse SQL and re-serialize it internally.

<img width="478" alt="image" src="https://github.com/ClickHouse/ClickHouse/assets/11692935/51d865e4-5a23-42f4-ab4e-b9b2be659988">

https://github.com/ClickHouse/ClickHouse/blob/330a20647037c3f9005801442936495ff79ba634/src/Parsers/examples/create_parser.cpp#L10-L23

But I don't know anything about C++ and its toolchain ...... 
I've tried a couple of times to bring up the relevant code individually for compilation, but without success 😂😂😂. 

If there are any official ideas to support this feature, I would be grateful. . . 😉

**Comment by vdimir** [comment]
To reformat a query, ClickHouse uses an in-memory AST representation, but it's just a C++ data structure and it's hard to use in foreign code. If you need only to reformat a query to a 'canonical' form as the client does, you may use the `clickhouse format` tool:

```bash
$ echo 'select name from system.numbers' | clickhouse format 
SELECT name
FROM system.numbers
$
```

Or just send some of the `EXPLAIN ...` queries that I mentioned.

Could you please tell me a bit more about the specific use case? What do you intend to do with a query, just reformat or programmatically modify it somehow?

**Comment by hungtcs** [comment]
Hello @vdimir 

Thanks for your reply.

I need to make changes to the input SQL by modifying the AST.

For instance, I control the number of output results by programmatically add a CTE in the outer layers of the input,

```sql
-- before 
select name from system.tables

-- after
with limit_wrapper as (
  select name from system.tables
)
select * from limit_wrapper limit 10
```

Also, I've modified the AST to include some data as a virtual table, for example:

```sql
-- input
select name, age from some_virtual_table

-- I will replace to
with some_virtual_table as (
  select data[1]::String as name, data[2]::UInt8 as age from (SELECT JSONExtractArrayRaw(arrayJoin(JSONExtractArrayRaw('[ ["name 1", 20],[ "name 2", 21] ]'))) as data) as arr  
)
select name, age from some_virtual_table
```

These are two examples of my requirements. 
In short, I need to programmatically modify the SQL entered by the user and then generate a new SQL to the Clickhouse server for execution.

Thanks.

**Reply by loserbird** [reply]
what i want to do is very similar to you. Have you found a good way to solve you problem?

**Reply by hungtcs** [reply]
> what i want to do is very similar to you. Have you found a good way to solve you problem?

I am currently using this library
https://github.com/AfterShip/clickhouse-sql-parser

**Comment by hungtcs** [comment]
I recently learned about Antlr, It seems that a parser can be built through Antlr.
And here's a fairly well-developed example of the grammar.

https://github.com/ClickHouse/ClickHouse/blob/master/utils/antlr/ClickHouseParser.g4

I created a repository here

https://github.com/hungtcs/clickhouse-sql-parser

**Comment by royxact** [comment]
Hi @hungtcs ,
Are you aware of this project? could be useful for your use case:
https://github.com/tobymao/sqlglot

**Reply by hungtcs** [reply]
> Hi @hungtcs , Are you aware of this project? could be useful for your use case: https://github.com/tobymao/sqlglot

Thanks @royxact , this is a super useful project.

**Comment by avaitla** [comment]
Another use case is similar to this: https://duckdb.org/docs/stable/data/json/sql_to_and_from_json.html

It's helpful for instance for taking an existing query and visualizing each cte to better understand the overall final resulting query.
